docs/gen-versions.py

```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import os
+import re
+import sys
+
+import git
+import git.exc
+import git.refs.tag
+
+# These are the tags that are present in git, but not in the original
+# versions.json file. We specify them here to ensure we don't create links to
+# documentation that doesn't exist, and to recreate part of the original
+# versions.json file.
+EXCLUDE_VERSIONS = ["0.26.5", "0.23.4", "0.23.3", "0.23.2", "0.23.1", "0.22.2", "0.22.1", "0.21.2", "0.21.1"]
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        prog="gen-versions.py",
+        description="Generate Sphinx version switcher JSON file from git tags.",
+    )
+
+    parser.add_argument("commit",
+        help="commit to stop at while walking the graph to look for tags, beyond which tags will be ignored",
+        metavar="commit",
+    )
+
+    parser.add_argument("-o", "--out-file",
+        help="path to output file, including filename, for generated versions JSON file",
+        metavar="path",
+        default=None,
+    )
+
+    # Escape-hatch argument to exclude versions that are returned from
+    # git-rev-list. This basically lets us recreate the original file we already
+    # have, along with appending subsequent new version tags. We need this
+    # because the git-rev-list tags returned are a superset of the versions
+    # originally in versions.json. I.e. some of the existing tagged patch
+    # release versions don't have corresponding separate doc links.
+    parser.add_argument("--exclude-versions",
+        help="comma-separated list of additional versions to exclude from the versions returned by walking the git DAG",
+        metavar="versions",
+        type=str,
+    )
+
+    return parser.parse_args()
+
+def main():
+    args = parse_args()
+
+    exclude_versions = []
+    if args.exclude_versions is not None:
+        # Include the default EXCLUDE_VERSIONS to reflect the historical
+        # versions.json file.
+        exclude_versions = args.exclude_versions.split(",")
+        exclude_versions.extend(EXCLUDE_VERSIONS)
+
+    # Probably run this from inside the repo somewhere.
+    try:
+        repo = git.Repo(os.getcwd(), search_parent_directories=True)
+    except git.exc.InvalidGitRepositoryError as e:
+        print("Invalid git repository: {}. Are you running this from a git repository?".format(e), file=sys.stderr)
```

Comment:
depending on your python version, you might be able to `e.add_note`

Reply:
this is preferred because it puts the note after the error

Reply:
As mentioned, unable to use `add_note` with Python 3.8. Resolving.